src/mcmini_private.cpp

```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
                                            initialTransition);
 }
 
-MC_PROGRAM_TYPE
-mc_run_initial_trace()
+// nextBranchPoint is an out parameter.
+void
+mc_explore_branch(MCOptional<int> *nextBranchPoint)
 {
-  MC_PROGRAM_TYPE program = mc_fork_new_trace_at_main();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  tid_t backtrackThread;
+
+  if (nextBranchPoint == nullptr /* initial branch */) {
+    mc_fork_new_trace();
+    backtrackThread = TID_MAIN_THREAD;
+  } else { // else next branch
+    const int bp = nextBranchPoint->unwrapped();
+    auto *sNext = &(programState->getStateItemAtIndex(bp));
+    backtrackThread = sNext->popThreadToBacktrackOn();
+
+    // Prepare the scheduler's model of the next trace
+    programState->reflectStateAtTransitionIndex(bp - 1);
+
+    mc_fork_next_trace_at_current_state();
+  }
 
-  mc_search_dpor_branch_with_initial_thread(TID_MAIN_THREAD);
+  mc_search_dpor_branch_with_thread(backtrackThread);
   mc_exit_with_trace_if_necessary(traceId);
-  program = mc_rerun_current_trace_as_needed();
+  mc_run_next_trace_for_debugger();
+
   traceId++;
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
-  return MC_SCHEDULER;
+  if (nextBranchPoint != nullptr) { // if this is not the initial branch
+    *nextBranchPoint = programState->getDeepestDPORBranchPoint();
+  }
```

Comment:
I'd move this `if` statement after what's now line 256 to give us something like:

```cpp
 while (nextBranchPoint.hasValue()) {
     MCOptional<int> *nextBranchPointPtr = &nextBranchPoint;
     mc_explore_branch(nextBranchPointPtr);
     *nextBranchPoint = programState->getDeepestDPORBranchPoint();
}
```

which will turn into (pending the other changes I suggested)
```cpp
 while (nextBranchPoint.hasValue()) {
     mc_explore_branch(nextBranchPoint.unwrapped());
     nextBranchPoint = programState->getDeepestDPORBranchPoint();
}
```
Moving the branch point over through modification of the pointer requires the reader to pick up on the fact that modification of the pointer of `nextBranchPointPtr` occurs inside this function. It is just a bit too subtle that's all: I  thought there was an issue on first glance at the changes made around the `while` loop

Reply:
As before, I think `nextBranchPointPtr != nullvalue` is clearer to a general reader of the code than `nextBranchPoint.hasValue()`.  Note also that GDB will not allow you to look inside `nextBranchPoint.hasValue()` to see what it means.  Let's keep the code easy for a general programmer, as opposed to a C++ enthusiast.